[20기 이승무] Mission 2 회원가입 엔드포인트

goback opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- 회원가입 엔드포인트 구현
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- user 모델 생성
- 이메일 또는 패스워드 누락시 {'message': 'KEY_ERROR'}, status code 400 반환
- 이메일 또는 패스워드 형식에 맞지 않을 시, {'message': 'EMAIL_ERROR'} 또는 {'message': 'PASSWORD_ERROR'}status=400 반환
- 이메일, 전화번호, 닉네임 중복시, {'message': 'ALREADY_EXIST_ERROR'}, status=400 반환
- POST 방식으로 전달받는 parameter 가 없을 시, {'message': 'EMPTY_ARGS_ERROR'}, status=400 반환
- 이외 알 수 없는 에러 발생 시, {'message': 'UNKNOWN_ERROR'}, status=400 반환
- 성공시, {'message': 'SUCCESS'}, status=201 반환

## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)
- 사용자 예외처리를 적용했습니다
- 유효성 검사 로직을 별도 클래스의 함수로 분리했는데, static 함수로 정의하는게 나았을까요?

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
조건을 만족한다면 각 빈칸에 x 표시 해주세요
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [X] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [X] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [X] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

Commit:
CREATE: User 모델 생성

Commit:
CREATE: 사용자 정의 Exception 생성
    customException.py

Commit:
CREATE: validations.py 생성
    User 모델관련 Validation(필수 입력항목 , 이메일 형식, 패스워드 형식)

Commit:
MODIFY: 회원가입
    - account/views.py

Commit:
CREATE: 회원가입 urls.py
    - westagram/urls.py
    - account/urls.py

Kang-kyunghun commented on students/seungmoo-lee/account/models.py:
```diff
@@ -1,3 +1,12 @@
 
-# Create your models here.
+class User(models.Model):                                                       # 회원 정보
+    email           = models.EmailField(max_length=100, unique=True)            # 이메일
```
각 기능마다 주석을 넣어 주어 어떤 기능을 나타내는지 잘 보여주었습니다. 주석 같은 경우 사람마다 스타일이 다르기 때문에 보통 회사에서는 통일된 스타일로 가는 경우가 많습니다. 위코드에서는 조금 더 깔끔하고, 주석에 대한 오해를 방지하고자 깃허브에 올라오는 코드에는 주석을 제거하고 있습니다. 주석 부분 제거 부탁드리겠습니다.

Kang-kyunghun commented on students/seungmoo-lee/account/validations.py:
```diff
@@ -0,0 +1,17 @@
+
+class UserValidation:   # User 모델관련 Validation
+
+    # 필수 입력항목 체크
```
이 파일도 주석은 제거 부탁 드립니다.

Kang-kyunghun commented on students/seungmoo-lee/account/validations.py:
```diff
@@ -0,0 +1,17 @@
+import re
```
validation 파일을 따로 빼셔서 관리하는 점 좋습니다~~!!

Kang-kyunghun commented on students/seungmoo-lee/account/views.py:
```diff
@@ -0,0 +1,49 @@
+from django.views         import View
+from django.db            import IntegrityError
+
+from .models                    import User
```
import 라인 맞춰주세요.

Kang-kyunghun commented on students/seungmoo-lee/account/views.py:
```diff
@@ -0,0 +1,49 @@
+        except Exception as e:
+            return JsonResponse({'message': 'UNKNOWN_ERROR'}, status=400)
+
+        return JsonResponse({'message': 'SUCCESS'}, status=201)
```
excrpt들을 가장 마지막에 두는 것이 가독성에 더 좋을 것 같습니다.!!

Kang-kyunghun commented on students/seungmoo-lee/account/views.py:
```diff
@@ -0,0 +1,49 @@
+            nickname     = data.get('nickname')
+
+            if userValidation.check_required_fields(email, password):
+                raise KeyError()
+            if userValidation.check_email(email):
+                raise ValidError('EMAIL_ERROR')
+            if userValidation.check_password(password):
+                raise ValidError('PASSWORD_ERROR')
```
raise를 사용하지 않고 바로 retrun하여 에러 처리하는 것이 가독성 면에서도 그렇고 코드 효율성면에서 떨어지는 것 같습니다.

Kang-kyunghun commented on students/seungmoo-lee/account/views.py:
```diff
@@ -0,0 +1,49 @@
+# 회원가입
+class SignUpView(View):
+    def post(self, request):
+        userValidation = UserValidation()
```
python의 변수명은 스네이크 케이스를 사용합니다.

Kang-kyunghun commented on students/seungmoo-lee/account/views.py:
```diff
@@ -0,0 +1,49 @@
+            return JsonResponse({'message': f'{e}'}, status=400)
+        except IntegrityError as e:
+            return JsonResponse({'message': 'ALREADY_EXIST_ERROR'}, status=400)
+        except Exception as e:
+            return JsonResponse({'message': 'UNKNOWN_ERROR'}, status=400)
```
예상치 못 한 에러를 제거하려고 작성한 것 같습니다. 그런데 이럴 경우 어떤 에러가 나는 것인지 정확히 디버깅 할 수 없습니다.

Kang-kyunghun commented on students/seungmoo-lee/account/views.py:
```diff
@@ -0,0 +1,49 @@
+            return JsonResponse({'message': 'KEY_ERROR'}, status=400)
+        except ValidError as e:
+            return JsonResponse({'message': f'{e}'}, status=400)
+        except IntegrityError as e:
```
이 예외 처리는 DB에 문제가 있을 겨우 다 처리 되는 예외 입니다. 그래서 명확한 예외 처리로 보기 힘들 것 같아요. 차라리 create 전에 데이터 존재 유무를 판단하는 것이 더 좋지 않을까 생각됩니다.

Commit:
MODIFY: 회원가입
    - account/models.py
        - 주석 제거
    - account/validations.py
        - 주석 제거
    - account/views.py
        -
    - exceptions/
        - 삭제

Commit:
MODIFY: 회원가입

goback commented:
수정 완료했습니다!

Kang-kyunghun commented on students/seungmoo-lee/account/views.py:
```diff
@@ -0,0 +1,48 @@
+
+            return JsonResponse({'message': 'SUCCESS'}, status=201)
+
+        except JSONDecodeError as e:
```
```suggestion
        except JSONDecodeError:
```
디버깅용 코드가 아니기 때문에 빼셔도 될 것 같습니다.

Kang-kyunghun commented on students/seungmoo-lee/account/views.py:
```diff
@@ -0,0 +1,48 @@
+            phone_number = data.get('phone_number')
+            nickname     = data.get('nickname')
+
+            if user_validation.check_required_fields(email, password):
+                return JsonResponse({'message': 'KEY_ERROR'}, status=400)
+            if user_validation.check_email(email):
+                return JsonResponse({'message': 'EMAIL_ERROR'}, status=400)
+            if user_validation.check_password(password):
+                return JsonResponse({'message': 'PASSWORD_ERROR'}, status=400)
```
```suggestion
            if user_validation.check_required_fields(email, password):
                return JsonResponse({'message': 'KEY_ERROR'}, status=400)
                
            if user_validation.check_email(email):
                return JsonResponse({'message': 'EMAIL_ERROR'}, status=400)
                
            if user_validation.check_password(password):
                return JsonResponse({'message': 'PASSWORD_ERROR'}, status=400)
```
가독성 측면에서 위와 같이 하는 것이 더 나을 것 같습니다.

Kang-kyunghun commented on students/seungmoo-lee/account/validations.py:
```diff
@@ -0,0 +1,14 @@
+        return email_pattern.match(email) is None
+
+    def check_password(self, password):
+        password_pattern = re.compile('^([^ ]{8,}$)+')
```
비밀번호 길이가 8보다 작을 때 조건을 잘 만드셨습니다.
그런데, 나중에 기획이 바뀌어서 길이가 10보다 작을 때로 바뀌면 숫자 부분을 바꿔야 합니다. 지금은 1부분에서만 사용 했기 때문에 쉽게 바꿀 수 있지만, 나중에 프로젝트가 커지고 저러한 부분이 많이 사용 되었다면, 한 번 수정 할 때 어려움이 있습니다. 그래서 이럴 경우에는 상수를 하나 만들어서 거기에 데이터를 저장해서 사용합니다.

Commit:
MODIFY: 회원가입
    - 수정사항 반영

Kang-kyunghun commented on students/seungmoo-lee/account/views.py:
```diff
@@ -0,0 +1,50 @@
+
+            return JsonResponse({'message': 'SUCCESS'}, status=201)
+
+        except JSONDecodeError:
+            return JsonResponse({'message': 'EMPTY_ARGS_ERROR'}, status=400)
```
해당 excrpt은 언제 발생 할까요? 코멘트로 남겨 주세요.

goback replied:
JSON 데이터를  파이썬 객체로 역직렬화 할 때, 역직렬화 대상이 되는 데이터가 올바른 JSON 데이터가 아닐 경우 발생해요.

Kang-kyunghun commented:
고생하셨습니다~~!!
